### PR TITLE
feature(invity): exchange crypto functionality

### PR DESCRIPTION
**Pull request by martinboehm**
This pull request adds exchange crypto functionality.

### CODE REVIEW

File: packages/components/src/components/form/Input/index.tsx [hunk 1]
```diff
@@ -150,6 +150,7 @@ interface Props extends React.InputHTMLAttributes<HTMLInputElement> {
     bottomText?: React.ReactNode;
     monospace?: boolean;
     isDisabled?: boolean;
+    className?: string;
```
Comment (szymonlesisz) [inline packages/components/src/components/form/Input/index.tsx:153]: unused prop

File: packages/suite/src/actions/wallet/coinmarket/coinmarketCommonActions.ts [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+    getState: GetState,
+) => {
+    const { device } = getState().suite;
+    if (device && !device.remember && !isDesktop()) {
+        dispatch(suiteActions.rememberDevice(device, true));
+    }
+    utilsSubmitRequestForm(tradeForm);
+};
+
+export const saveComposedTransaction = (composedTransaction: PrecomposedTransactionFinal) => async (
```
Comment (szymonlesisz) [inline packages/suite/src/actions/wallet/coinmarket/coinmarketCommonActions.ts]: unnecessary thunk action, use regular redux action:
`export const saveComposedTransaction = (...params): CoinmarketCommonAction => ({ type, ...rest });`

File: packages/suite/src/actions/wallet/coinmarket/coinmarketCommonActions.ts [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+    dispatch: Dispatch,
+) => {
+    dispatch({
+        type: COINMARKET_COMMON.SAVE_COMPOSED_TRANSACTION,
+        composedTransaction,
+    });
+};
+
+export const saveTransactionReview = (reviewData: ReviewTransactionData) => async (
```
Comment (szymonlesisz) [inline packages/suite/src/actions/wallet/coinmarket/coinmarketCommonActions.ts]: unnecessary thunk action, use regular redux action

File: packages/suite/src/actions/wallet/coinmarketExchangeActions.ts [hunk 4]
```diff
@@ -32,46 +39,88 @@ export type CoinmarketExchangeAction =
...
               symbol: Account['symbol'];
+              descriptor: Account['descriptor'];
               accountIndex: Account['index'];
               accountType: Account['accountType'];
           };
       };
 
-export const loadExchangeInfo = async (): Promise<ExchangeInfo> => {
-    const exchangeList = await invityAPI.getExchangeList();
+export async function loadExchangeInfo(): Promise<[ExchangeInfo, ExchangeCoinInfo[]]> {
```
Comment (szymonlesisz) [inline packages/suite/src/actions/wallet/coinmarketExchangeActions.ts]: why changing const for function? lets be consistent

File: packages/suite/src/actions/wallet/coinmarketExchangeActions.ts [hunk 5]
```diff
@@ -32,46 +39,88 @@ export type CoinmarketExchangeAction =
...
+            exchangeList,
+            providerInfos,
+            buySymbols,
+            sellSymbols,
+        },
+        exchangeCoinInfo,
+    ];
+}
+
+export const saveExchangeInfo = (exchangeInfo: ExchangeInfo) => async (dispatch: Dispatch) => {
```
Comment (szymonlesisz) [inline packages/suite/src/actions/wallet/coinmarketExchangeActions.ts]: unnecessary thunk action, use regular redux action

File: packages/suite/src/actions/wallet/coinmarketExchangeActions.ts [hunk 6]
```diff
@@ -32,46 +39,88 @@ export type CoinmarketExchangeAction =
...
+    dispatch({
+        type: COINMARKET_EXCHANGE.SAVE_EXCHANGE_INFO,
+        exchangeInfo,
+    });
+};
+
+export const saveExchangeCoinInfo = (exchangeCoinInfo: ExchangeCoinInfo[]) => async (
```
Comment (szymonlesisz) [inline packages/suite/src/actions/wallet/coinmarketExchangeActions.ts]: unnecessary thunk action, use regular redux action

File: packages/suite/src/components/suite/QuestionTooltip/index.tsx [hunk 7]
```diff
@@ -17,6 +17,7 @@ interface Props {
     label?: JSX.Element | ExtendedMessageDescriptor['id'];
     tooltip?: JSX.Element | ExtendedMessageDescriptor['id'];
     size?: number;
+    className?: string;
```
Comment (szymonlesisz) [inline packages/suite/src/components/suite/QuestionTooltip/index.tsx:20]: unused prop

File: packages/suite/src/components/suite/modals/confirm/CoinmarketExchangeTerms/index.tsx [hunk 8]
```diff
@@ -0,0 +1,130 @@
+import { Button, Modal, Icon, colors, variables, Checkbox } from '@trezor/components';
+import React, { useState } from 'react';
+import { Translation } from '@suite-components';
+import styled, { css } from 'styled-components';
+import { Deferred } from '@suite/utils/suite/deferred';
```
Comment (szymonlesisz) [inline packages/suite/src/components/suite/modals/confirm/CoinmarketExchangeTerms/index.tsx]: from `@suite-utils/deferred`

File: packages/suite/src/components/suite/modals/index.tsx [hunk 9]
```diff
@@ -98,8 +102,13 @@ const getDeviceContextModal = (props: Props) => {
         case 'ButtonRequest_PinEntry':
             return <ConfirmAction device={device} />;
         case 'ButtonRequest_ConfirmOutput':
-        case 'ButtonRequest_SignTx':
+        case 'ButtonRequest_SignTx': {
+            if (isCoinmarketOffers) {
+                return <CoinmarketReviewTransaction type="sign-transaction" />;
```
Comment (szymonlesisz) [inline packages/suite/src/components/suite/modals/index.tsx:107]: i dont like this duplicates, same as duplicated signing actions

why they cant be shared?

File: packages/suite/src/components/wallet/CoinmarketLayout/components/AccountTransactions/components/ExchangeTransaction/index.tsx [hunk 10]
```diff
@@ -1,15 +1,24 @@
...
+import { Button, colors, Icon, variables } from '@trezor/components';
+import { CoinmarketExchangeProviderInfo } from '@wallet-components';
 import { TradeExchange } from '@wallet-reducers/coinmarketReducer';
-import { formatDistance } from 'date-fns';
+import * as routerActions from '@suite-actions/routerActions';
+import * as coinmarketExchangeActions from '@wallet-actions/coinmarketExchangeActions';
+import { Account } from '@wallet-types';
+import { useWatchExchangeTrade } from '@wallet-hooks/useCoinmarket';
+import Status from '../Status';
+import { Translation } from '@suite/components/suite';
```
Comment (szymonlesisz) [inline packages/suite/src/components/wallet/CoinmarketLayout/components/AccountTransactions/components/ExchangeTransaction/index.tsx]: from `@suite-components`

File: packages/suite/src/components/wallet/CoinmarketLayout/components/AccountTransactions/components/ExchangeTransaction/index.tsx [hunk 11]
```diff
@@ -1,15 +1,24 @@
...
+import { useActions } from '@suite/hooks/suite';
```
Comment (szymonlesisz) [inline packages/suite/src/components/wallet/CoinmarketLayout/components/AccountTransactions/components/ExchangeTransaction/index.tsx]: from `@suite-hooks`

File: packages/suite/src/components/wallet/CoinmarketLayout/components/AccountTransactions/components/ExchangeTransaction/index.tsx [hunk 12]
```diff
@@ -47,41 +87,87 @@ const Row = styled.div`
...
+
+const Arrow = styled.div`
+    display: flex;
+    align-items: center;
+    padding: 0 11px;
 `;
 
-const ExchangeTransaction = ({ trade, providers }: Props) => {
+const ExchangeTransaction = ({ trade, providers, account }: Props) => {
+    const { goto } = useActions({ goto: routerActions.goto });
```
Comment (szymonlesisz) [inline packages/suite/src/components/wallet/CoinmarketLayout/components/AccountTransactions/components/ExchangeTransaction/index.tsx]: add goto: to the block below (unnecessary double usage of useAction)

File: packages/suite/src/hooks/wallet/useCoinmarketBuyOffers.ts [hunk 13]
```diff
@@ -3,7 +3,7 @@ import invityAPI from '@suite-services/invityAPI';
 import { useActions, useSelector } from '@suite-hooks';
 import { BuyTrade } from 'invity-api';
 import { processQuotes } from '@wallet-utils/coinmarket/buyUtils';
-import * as coinmarketCommonActions from '@wallet-actions/coinmarketCommonActions';
+import * as coinmarketCommonActions from '@wallet-actions/coinmarket/coinmarketCommonActions';
 import * as coinmarketBuyActions from '@wallet-actions/coinmarketBuyActions';
 import * as routerActions from '@suite-actions/routerActions';
 import { createQuoteLink, createTxLink } from '@suite/utils/wallet/coinmarket/buyUtils';
```
Comment (szymonlesisz) [inline packages/suite/src/hooks/wallet/useCoinmarketBuyOffers.ts]: merge this line with line 5 (double import from the same file)

File: packages/suite/src/hooks/wallet/useCoinmarketExchangeForm.ts [hunk 14]
```diff
@@ -0,0 +1,282 @@
...
+    Props,
+    ComposeData,
+    AmountLimits,
+    ExchangeFormContextValues,
+} from '@wallet-types/coinmarketExchangeForm';
+import {
+    getAmountLimits,
+    splitToFixedFloatQuotes,
+} from '@suite/utils/wallet/coinmarket/exchangeUtils';
+import { ExchangeTradeQuoteRequest } from 'invity-api';
```
Comment (szymonlesisz) [inline packages/suite/src/hooks/wallet/useCoinmarketExchangeForm.ts]: pls reoreder this imports, its really hard to read

File: packages/suite/src/types/wallet/coinmarketExchangeOffers.ts [hunk 15]
```diff
@@ -0,0 +1,51 @@
+import { AppState } from '@suite-types';
+import { Account } from '@wallet-types';
+import { ExchangeTrade } from 'invity-api';
+import {
+    CoinmarketExchangeAction,
+    ExchangeInfo,
+} from '@suite/actions/wallet/coinmarketExchangeActions';
```
Comment (szymonlesisz) [inline packages/suite/src/types/wallet/coinmarketExchangeOffers.ts]: from `@wallet-actions`

File: packages/suite/src/utils/wallet/coinmarket/exchangeUtils.ts [hunk 16]
```diff
@@ -0,0 +1,132 @@
+import { ExchangeInfo } from '@wallet-actions/coinmarketExchangeActions';
+import { AmountLimits } from '@wallet-types/coinmarketExchangeForm';
+import { ExchangeTrade, ExchangeTradeStatus } from 'invity-api';
+import { symbolToInvityApiSymbol } from './coinmarketUtils';
+import { Account } from '@wallet-types';
+
+// loop through quotes and if all quotes are either with error below minimum or over maximum, return error message
+export function getAmountLimits(quotes: ExchangeTrade[]): AmountLimits | undefined {
```
Comment (szymonlesisz) [inline packages/suite/src/utils/wallet/coinmarket/exchangeUtils.ts]: use const, lets be consistent

File: packages/suite/src/utils/wallet/coinmarket/exchangeUtils.ts [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+        if (!quote.error && noError) {
+            return;
+        }
+    }
+    if (min || max) {
+        return { currency: quotes[0].send || '', min, max };
+    }
+}
+
+export function isQuoteError(quote: ExchangeTrade): boolean {
```
Comment (szymonlesisz) [inline packages/suite/src/utils/wallet/coinmarket/exchangeUtils.ts]: use const, lets be consistent + unnecessary return type

File: packages/suite/src/utils/wallet/coinmarket/exchangeUtils.ts [hunk 18]
```diff
@@ -0,0 +1,132 @@
...
+        return true;
+    }
+    if (quote.max && quote.max !== 'NONE' && Number(quote.sendStringAmount) > quote.max) {
+        return true;
+    }
+    return false;
+}
+
+// return 3 arrays: quotes not in error, quotes with min/max error, quotes with general error
+function splitQuotes(quotes: ExchangeTrade[]): [ExchangeTrade[], ExchangeTrade[], ExchangeTrade[]] {
```
Comment (szymonlesisz) [inline packages/suite/src/utils/wallet/coinmarket/exchangeUtils.ts]: use const, lets be consistent

File: packages/suite/src/utils/wallet/coinmarket/exchangeUtils.ts [hunk 19]
```diff
@@ -0,0 +1,132 @@
...
+    return [
+        quotes.filter(q => !isQuoteError(q)),
+        quotes.filter(q => isQuoteError(q) && !q.error),
+        quotes.filter(q => q.error),
+    ];
+}
+
+export function splitToFixedFloatQuotes(
```
Comment (szymonlesisz) [inline packages/suite/src/utils/wallet/coinmarket/exchangeUtils.ts]: use const, lets be consistent

File: packages/suite/src/views/wallet/coinmarket/exchange/components/ExchangeForm/Footer/index.tsx [hunk 20]
```diff
@@ -0,0 +1,44 @@
+import React from 'react';
+import { Button } from '@trezor/components';
+import { Translation } from '@suite-components';
+import styled from 'styled-components';
+import { useCoinmarketExchangeFormContext } from '@suite/hooks/wallet/useCoinmarketExchangeForm';
```
Comment (szymonlesisz) [inline packages/suite/src/views/wallet/coinmarket/exchange/components/ExchangeForm/Footer/index.tsx]: from `@wallet-hooks`

File: packages/suite/src/views/wallet/coinmarket/exchange/components/ExchangeForm/Inputs/Buttons/index.tsx [hunk 21]
```diff
@@ -0,0 +1,112 @@
+import { colors, variables } from '@trezor/components';
+import React from 'react';
+import BigNumber from 'bignumber.js';
+import { Translation } from '@suite-components';
+import styled from 'styled-components';
+import { invityApiSymbolToSymbol } from '@wallet-utils/coinmarket/coinmarketUtils';
+import { useCoinmarketExchangeFormContext } from '@suite/hooks/wallet/useCoinmarketExchangeForm';
```
Comment (szymonlesisz) [inline packages/suite/src/views/wallet/coinmarket/exchange/components/ExchangeForm/Inputs/Buttons/index.tsx]: from `@wallet-hooks`

File: packages/suite/src/views/wallet/coinmarket/exchange/components/ExchangeForm/Inputs/FiatInput/index.tsx [hunk 22]
```diff
@@ -0,0 +1,71 @@
+import { Input } from '@trezor/components';
+import React from 'react';
+import styled from 'styled-components';
+import { InputError } from '@wallet-components';
+import { isDecimalsValid } from '@wallet-utils/validation';
+import { useCoinmarketExchangeFormContext } from '@suite/hooks/wallet/useCoinmarketExchangeForm';
```
Comment (szymonlesisz) [inline packages/suite/src/views/wallet/coinmarket/exchange/components/ExchangeForm/Inputs/FiatInput/index.tsx]: from `@wallet-hooks`

File: packages/suite/src/views/wallet/coinmarket/exchange/components/ExchangeForm/Inputs/FiatInput/index.tsx [hunk 23]
```diff
@@ -0,0 +1,71 @@
...
+import { Translation } from '@suite/components/suite';
```
Comment (szymonlesisz) [inline packages/suite/src/views/wallet/coinmarket/exchange/components/ExchangeForm/Inputs/FiatInput/index.tsx]: from `@suite-components`

File: packages/suite/src/views/wallet/coinmarket/exchange/components/ExchangeForm/Inputs/SellCryptoSelect/index.tsx [hunk 24]
```diff
@@ -0,0 +1,154 @@
+import { variables, Select, colors } from '@trezor/components';
+import { ExchangeInfo } from '@wallet-actions/coinmarketExchangeActions';
+import React from 'react';
+import { Controller } from 'react-hook-form';
+import styled from 'styled-components';
+import { ExchangeCoinInfo } from 'invity-api';
+import { useCoinmarketExchangeFormContext } from '@suite/hooks/wallet/useCoinmarketExchangeForm';
```
Comment (szymonlesisz) [inline packages/suite/src/views/wallet/coinmarket/exchange/components/ExchangeForm/Inputs/SellCryptoSelect/index.tsx]: from `@wallet-hooks`

File: packages/suite/src/views/wallet/coinmarket/exchange/components/ExchangeForm/Inputs/index.tsx [hunk 25]
```diff
@@ -0,0 +1,81 @@
+import { Icon, variables } from '@trezor/components';
+import React, { useEffect } from 'react';
+import styled from 'styled-components';
+import { invityApiSymbolToSymbol } from '@wallet-utils/coinmarket/coinmarketUtils';
+import { useCoinmarketExchangeFormContext } from '@suite/hooks/wallet/useCoinmarketExchangeForm';
```
Comment (szymonlesisz) [inline packages/suite/src/views/wallet/coinmarket/exchange/components/ExchangeForm/Inputs/index.tsx]: from `@wallet-hooks`

**Comment by tsusanka** [comment]
<s>Please squash into one commit as we've agreed. There is no added value in filing PR with 153 commits if we squash at the end anyway.</s>

Discussed on Slack, that this is just a draft and was agreed upon beforehand.

File: packages/suite/src/views/wallet/coinmarket/exchange/components/ExchangeForm/index.tsx [hunk 26]
```diff
@@ -0,0 +1,67 @@
+import React from 'react';
+import styled from 'styled-components';
+import { variables, colors } from '@trezor/components';
+import { Translation } from '@suite-components';
+import { useCoinmarketExchangeFormContext } from '@suite/hooks/wallet/useCoinmarketExchangeForm';
```
Comment (szymonlesisz) [inline packages/suite/src/views/wallet/coinmarket/exchange/components/ExchangeForm/index.tsx]: from `@wallet-hooks`

**Review by szymonlesisz — CHANGES_REQUESTED** [review]
i dont like duplicates of ReviewTransaction and sendActions, lets think how to make it shareable

second thing it this autoimport, pls disable it for this project. it's really waste of time for me and you do keep focusing and fixing that all over again.

third thing is a gulash in imports in general, it's really hard to read. Imagine yourself to read this kinda pullrequest with such a mess

File: packages/components/src/components/form/Input/index.tsx (hunk 1 above)
Comment (martinboehm) [reply packages/components/src/components/form/Input/index.tsx:153]: It is used for styled components.

File: packages/suite/src/components/suite/QuestionTooltip/index.tsx (hunk 7 above)
Comment (martinboehm) [reply packages/suite/src/components/suite/QuestionTooltip/index.tsx:20]: It is used for styled components.

File: packages/components/src/components/form/Input/index.tsx (hunk 1 above)
Comment (szymonlesisz) [reply packages/components/src/components/form/Input/index.tsx:153]: could you point me to implementation then? because i couldn't find any

File: packages/suite/src/components/suite/QuestionTooltip/index.tsx (hunk 7 above)
Comment (szymonlesisz) [reply packages/suite/src/components/suite/QuestionTooltip/index.tsx:20]: could you point me to implementation then? because i couldn't find any

**Review by szymonlesisz — CHANGES_REQUESTED** [review]

File: packages/components/src/components/form/Input/index.tsx (hunk 1 above)
Comment (vladimirvolek) [reply packages/components/src/components/form/Input/index.tsx:153]: This is how styled-components work.
Please see https://styled-components.com/docs/basics#styling-any-component

image

Comment (szymonlesisz) [reply packages/components/src/components/form/Input/index.tsx:153]: so how it's possible that it currently works without this prop? for example in PassphraseInput or other styled inputs not related to this PR?

File: packages/suite/src/components/suite/QuestionTooltip/index.tsx (hunk 7 above)
Comment (martinboehm) [reply packages/suite/src/components/suite/QuestionTooltip/index.tsx:20]: We use it here: https://github.com/trezor/trezor-suite/blob/9365e5388e60c9406c629763f34da28c00f0d183/packages/suite/src/views/wallet/coinmarket/exchange/offers/Offers/List/Quote/index.tsx#L108

File: packages/suite/src/views/wallet/coinmarket/exchange/offers/Offers/SelectedOffer/components/VerifyAddress/index.tsx [hunk 27]
```diff
@@ -0,0 +1,462 @@
...
+    Icon,
+    Button,
+} from '@trezor/components';
+import { InputError } from '@wallet-components';
+import { useCoinmarketExchangeOffersContext } from '@wallet-hooks/useCoinmarketExchangeOffers';
+import { Account } from '@wallet-types';
+import * as modalActions from '@suite-actions/modalActions';
+import { useDispatch } from 'react-redux';
+import { Dispatch } from '@suite-types';
+import { useTimeoutFn } from 'react-use';
```
Comment (szymonlesisz) [inline packages/suite/src/views/wallet/coinmarket/exchange/offers/Offers/SelectedOffer/components/VerifyAddress/index.tsx]: `import useTimeoutFn from 'react-use/lib/useTimeoutFn';`
this was already [refactored once](https://github.com/trezor/trezor-suite/commit/44e4da0d8150da6d4033875eecfccb63c227dbab) you want to import only relevant module not the whole lib

File: packages/suite-native/src/utils/wallet/coinmarket/buyUtils.ts [hunk 28]
```diff
@@ -0,0 +1,5 @@
+export function createQuoteLink() {}
+export function createTxLink() {}
+export function submitRequestForm(arg: any) {
+    return arg;
+}
```
Comment (szymonlesisz) [inline packages/suite-native/src/utils/wallet/coinmarket/buyUtils.ts]: for discussion, why is this here? this should be avoided

File: packages/suite/src/actions/wallet/coinmarket/coinmarketCommonActions.ts [hunk 29]
```diff
@@ -0,0 +1,260 @@
...
+                ...modalPayload,
+            }),
+        );
+        return;
+    }
+
+    const params = {
+        device,
+        path,
+        useEmptyPassphrase,
```
Comment (szymonlesisz) [inline packages/suite/src/actions/wallet/coinmarket/coinmarketCommonActions.ts:64]: missing param coin: symbol, [recently fixed](https://github.com/trezor/trezor-suite/commit/96cdf0e0bad89e13a39bc0c13e84b461ba1d5e8b)

File: packages/suite/src/views/wallet/coinmarket/exchange/offers/index.tsx [hunk 30]
```diff
@@ -0,0 +1,50 @@
...
+
+const OffersIndex = (props: ComponentProps) => {
+    const { selectedAccount } = props;
+    if (selectedAccount.status !== 'loaded') {
+        return <WalletLayout title="TR_NAV_EXCHANGE" account={selectedAccount} />;
+    }
+    return <OffersIndexLoaded {...props} selectedAccount={selectedAccount} />;
+};
+
+export default connect(mapStateToProps)(OffersIndex);
```
Comment (szymonlesisz) [inline packages/suite/src/views/wallet/coinmarket/exchange/offers/index.tsx]: lets use  hooks instead of connected component from now on, like it was [changed here](https://github.com/trezor/trezor-suite/commit/2fa02c3e3711c6c9c1cb9019d784eb4cf5d86942)

you don't have to introduce children props, but it would be useful to write hooks tests

File: packages/suite/src/views/wallet/coinmarket/exchange/index.tsx [hunk 31]
```diff
@@ -1,8 +1,47 @@
...
+    const { selectedAccount } = props;
+    if (selectedAccount.status !== 'loaded') {
+        return <WalletLayout title="TR_NAV_EXCHANGE" account={selectedAccount} />;
+    }
+
+    return <CoinmarketExchangeLoaded {...props} selectedAccount={selectedAccount} />;
 };
 
-export default CoinmarketBuy;
+export default connect(mapStateToProps)(CoinmarketExchange);
```
Comment (szymonlesisz) [inline packages/suite/src/views/wallet/coinmarket/exchange/index.tsx]: lets use hooks instead of connected component from now on, like it was [changed here](https://github.com/trezor/trezor-suite/commit/2fa02c3e3711c6c9c1cb9019d784eb4cf5d86942)

you don't have to introduce children props, but it would be useful to write hooks tests

File: packages/suite/src/views/wallet/coinmarket/exchange/detail/index.tsx [hunk 32]
```diff
@@ -0,0 +1,51 @@
...
+const DetailIndex = (props: ComponentProps) => {
+    const { selectedAccount } = props;
+
+    if (selectedAccount.status !== 'loaded') {
+        return <WalletLayout title="TR_NAV_EXCHANGE" account={selectedAccount} />;
+    }
+    return <DetailIndexLoaded {...props} selectedAccount={selectedAccount} />;
+};
+
+export default connect(mapStateToProps)(DetailIndex);
```
Comment (szymonlesisz) [inline packages/suite/src/views/wallet/coinmarket/exchange/detail/index.tsx]: lets use hooks instead of connected component from now on, like it was [changed here](https://github.com/trezor/trezor-suite/commit/2fa02c3e3711c6c9c1cb9019d784eb4cf5d86942)

you don't have to introduce children props, but it would be useful to write hooks tests

File: packages/suite/src/views/wallet/coinmarket/exchange/components/ExchangeForm/Inputs/ReceiveCryptoInput/index.tsx [hunk 33]
```diff
@@ -0,0 +1,165 @@
+import { Input } from '@trezor/components';
+import React from 'react';
+import { formatNetworkAmount } from '@wallet-utils/accountUtils';
+import { FIAT } from '@suite-config';
+import styled from 'styled-components';
+import { useDebounce } from 'react-use';
```
Comment (szymonlesisz) [inline packages/suite/src/views/wallet/coinmarket/exchange/components/ExchangeForm/Inputs/ReceiveCryptoInput/index.tsx]: `import useDebounce from 'react-use/lib/useDebounce';`
this was already [refactored once](https://github.com/trezor/trezor-suite/commit/44e4da0d8150da6d4033875eecfccb63c227dbab) you want to import only relevant module not the whole lib

File: packages/suite/src/hooks/wallet/useCoinmarketBuyDetail.ts [hunk 34]
```diff
@@ -25,7 +25,7 @@ export const useCoinmarketBuyDetail = (props: Props) => {
 
     return {
         account,
-        trade: buyTrade,
+        trade: buyTrade as TradeBuy,
```
Comment (szymonlesisz) [inline packages/suite/src/hooks/wallet/useCoinmarketBuyDetail.ts]: to discuss: why?

**Review by szymonlesisz — CHANGES_REQUESTED** [review]

File: packages/suite/src/views/wallet/coinmarket/exchange/offers/Offers/SelectedOffer/components/VerifyAddress/index.tsx (hunk 27 above)
Comment (martinboehm) [reply packages/suite/src/views/wallet/coinmarket/exchange/offers/Offers/SelectedOffer/components/VerifyAddress/index.tsx]: Done

File: packages/suite/src/actions/wallet/coinmarket/coinmarketCommonActions.ts (hunk 29 above)
Comment (martinboehm) [reply packages/suite/src/actions/wallet/coinmarket/coinmarketCommonActions.ts:64]: Done

File: packages/suite/src/views/wallet/coinmarket/exchange/offers/index.tsx (hunk 30 above)
Comment (martinboehm) [reply packages/suite/src/views/wallet/coinmarket/exchange/offers/index.tsx]: Done

File: packages/suite/src/views/wallet/coinmarket/exchange/index.tsx (hunk 31 above)
Comment (martinboehm) [reply packages/suite/src/views/wallet/coinmarket/exchange/index.tsx]: Done, for now without children

File: packages/suite/src/views/wallet/coinmarket/exchange/detail/index.tsx (hunk 32 above)
Comment (martinboehm) [reply packages/suite/src/views/wallet/coinmarket/exchange/detail/index.tsx]: Done, for now without children

File: packages/suite/src/views/wallet/coinmarket/exchange/components/ExchangeForm/Inputs/ReceiveCryptoInput/index.tsx (hunk 33 above)
Comment (martinboehm) [reply packages/suite/src/views/wallet/coinmarket/exchange/components/ExchangeForm/Inputs/ReceiveCryptoInput/index.tsx]: Done

File: packages/suite/src/hooks/wallet/useCoinmarketBuyDetail.ts (hunk 34 above)
Comment (martinboehm) [reply packages/suite/src/hooks/wallet/useCoinmarketBuyDetail.ts]: Done

File: packages/suite-native/src/utils/wallet/coinmarket/buyUtils.ts (hunk 28 above)
Comment (martinboehm) [reply packages/suite-native/src/utils/wallet/coinmarket/buyUtils.ts]: The reason is that the functionality is now referenced from coinmarketCommonActions. However,
we changed the implementation a bit - all web/destop specific code is now in `dom.ts`, with native code in native `dom.ts`, for now only stubs functions.

Comment (szymonlesisz) [reply packages/suite-native/src/utils/wallet/coinmarket/buyUtils.ts]: so now it's even worst. having DOM in react-native is a oxymoron :), move it pls to `/utils/env`

Comment (martinboehm) [reply packages/suite-native/src/utils/wallet/coinmarket/buyUtils.ts]: Done

**Review by keraf — APPROVED** [review]
Looks good! There are just a few of the eslint exceptions that could be addressed but I'm fine with this being done at a later time (we anyway need to check the whole project for these at some point).

File: packages/components/src/components/form/Input/index.tsx (hunk 1 above)
Comment (szymonlesisz) [reply packages/components/src/components/form/Input/index.tsx:153]: im still not conviced :)

**Review by szymonlesisz — APPROVED** [review]
